### ISSUE TITLE
How Do I fine-tune a custom DETR model as we fine-tune Hugging face transformer

### ISSUE THREAD

**Issue by tanulsingh** [facebookresearch/detr#109]
I am doing something like

` 
class CustomDETR(nn.Module):

    def __init__(self):
        super(CustomDETR,self).__init__()
        self.model = torch.hub.load('facebookresearch/detr', 'detr_resnet50', pretrained=True)
        self.model.class_embed = nn.Linear(in_features=256,out_features=2)
        self.model.num_queries = 100
    
    def forward(self,):
        output = self.model()
 `

My question is the output of the model is class and bboxes and no loss , how do I calculate the loss and backpropagate it so that I can fine tune the model

**Comment by fmassa** [comment]
Hi,

There is a lot of useful information about finetuning in https://github.com/facebookresearch/detr/issues/9, so let's keep the discussion there.

For your particular question, you should also add a `SetCriterion`, which takes the predictions and the ground-truths and outputs a loss, see https://github.com/facebookresearch/detr/blob/10a2c759454930813aeac7af5e779f835dcb75f5/models/detr.py#L83-L88

Our training loop looks like the following https://github.com/facebookresearch/detr/blob/10a2c759454930813aeac7af5e779f835dcb75f5/engine.py#L32-L33

IMO, the easiest way to get started with training is to clone the repo and modify the functions you need, see https://github.com/facebookresearch/detr#training for information.

I hope this helps.

I believe I've answered your question, and as such I'm closing this issue, but let us know if you have further questions

**Comment by tanulsingh** [comment]
@fmassa I was able to import HungarianMatcher from detr.models.matcher and it was working fine , but now it says module util not found after the last update I guess , I wrote a whole pipeline and somehow now it doesn't work , I would really appreciate some help

**Comment by fmassa** [comment]
@tanulsingh there should be no breaking change in the repo since its release. One thing to keep in mind is that you need to run the code from the repo root, as DETR is not packaged as a library

**Comment by tanulsingh** [comment]
@fmassa  Thanks It helped I was able to write a complete pipeline that can be used to train DETR on any custom dataset just by changing the dataloader 
https://www.kaggle.com/tanulsingh077/end-to-end-object-detection-with-transformers-detr
I would be glad if you share your views/suggestions on it , that I can improve
Thanks a lot
